### PR TITLE
Correctly concatenate pandas.Index objects

### PR DESCRIPTION
**Pull request by shoyer**
We now have a dedicated method, Coordinate.concat that will correctly
concatenate pandas.Index objects, even if they can't be properly expressed as
NumPy arrays (e.g., PeriodIndex and MultiIndex).

As part of this change, I removed and replaced the internal `interleaved_concat`
routine. It turns out we can do this with an inverse permutation instead, which
results in much simpler and cleaner code. In particular, we no longer need a
special path to support dask.array.

This should help with #818.

### CODE REVIEW

File: xarray/core/groupby.py [hunk 1]
```diff
@@ -66,6 +67,52 @@ def _dummy_copy(xarray_obj):
         raise AssertionError
     return res
 
+def _is_one_or_none(obj):
+    return obj == 1 or obj is None
```
Comment (jhamman) [inline xarray/core/groupby.py]: do you want to treat `True` as `1`?  If not, this should be `obj is 1 or obj is None`.

Comment (MaximilianR) [inline xarray/core/groupby.py]: Interesting! I never knew that. 
I always thought ints & strings should use `==`, but for -1 to 256 `is` is fine: http://stackoverflow.com/questions/2239737/is-it-better-to-use-is-or-for-number-comparison-in-python

Comment (fmaussion) [inline xarray/core/groupby.py]: But this depends on the interpreter, then. CPython might work, but others might not I guess?

Comment (shoyer) [inline xarray/core/groupby.py]: These slice objects are generated internally by groupby, so I think it's pretty far fetched to get `step is True`. Using `==` feels more idiomatic, anyways.

File: xarray/core/variable.py [hunk 2]
```diff
@@ -1163,3 +1208,41 @@ def _broadcast_compat_data(self, other):
...
+        arrays to concatenate are skipped.
+
+    Returns
+    -------
+    stacked : Variable
+        Concatenated Variable formed by stacking all the supplied variables
+        along the given dimension.
+    """
+    variables = list(variables)
+    print(variables)
```
Comment (MaximilianR) [inline xarray/core/variable.py]: guessing this is left over from debugging

Comment (shoyer) [inline xarray/core/variable.py]: fixed, thanks

**Pull request merged**